source/postprocess/rotation_statistics.cc

```diff
@@ -0,0 +1,260 @@
+
+      MaterialModel::MaterialModelInputs<dim> in(fe_values.n_quadrature_points, this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, this->n_compositional_fields());
+
+      Tensor<1,dim> local_angular_momentum;
+      double local_scalar_angular_momentum = 0.0;
+
+      SymmetricTensor<2,dim> local_moment_of_inertia;
+      double local_scalar_moment_of_inertia = 0.0;
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+      for (; cell!=endc; ++cell)
+        if (cell->is_locally_owned())
+          {
+            fe_values.reinit (cell);
+
+            // Set use_strain_rates to false since we don't need viscosity
+            in.reinit(fe_values, cell, this->introspection(), this->get_solution(), false);
+            if (!use_constant_density)
+              {
+                this->get_material_model().evaluate(in, out);
+              }
+
+            for (unsigned int q = 0; q < fe_values.n_quadrature_points; ++q)
+              {
+                const double rho = (use_constant_density)
+                                   ?
+                                   1.0
+                                   :
+                                   out.densities[q];
+
+                if (dim == 2)
+                  {
+                    // Get the velocity perpendicular to the position vector
+                    const Tensor<1,dim> r_perp = cross_product_2d(in.position[q]);
+                    // calculate a signed scalar angular momentum
+                    local_scalar_angular_momentum += in.velocity[q] * r_perp * rho * fe_values.JxW(q);
+                    local_scalar_moment_of_inertia += in.position[q].norm_square() * rho * fe_values.JxW(q);
+                  }
+                else if (dim == 3)
+                  {
+                    const Point<dim> position = in.position[q];
+                    const Tensor<1,dim> r_cross_v = cross_product_3d(in.position[q], in.velocity[q]);
+
+                    local_angular_momentum += r_cross_v * rho * fe_values.JxW(q);
+
+                    // calculate moment of inertia
+                    local_moment_of_inertia[0][0] += (position[1] * position[1] + position[2] * position[2]) * rho * fe_values.JxW(q);
+                    local_moment_of_inertia[1][1] += (position[0] * position[0] + position[2] * position[2]) * rho * fe_values.JxW(q);
+                    local_moment_of_inertia[2][2] += (position[0] * position[0] + position[1] * position[1]) * rho * fe_values.JxW(q);
+                    local_moment_of_inertia[0][1] -= (position[0] * position[1]) * rho * fe_values.JxW(q);
+                    local_moment_of_inertia[0][2] -= (position[0] * position[2]) * rho * fe_values.JxW(q);
+                    local_moment_of_inertia[1][2] -= (position[1] * position[2]) * rho * fe_values.JxW(q);
```

Comment:
I bet one could have done that in a dimension-independent way with only one line. Aren't these all just some outer  products and tensor products?

Reply:
Not that it really matters, though. So feel free to ignore my comment.